hipogram/posts/models.py

```diff
@@ -13,8 +15,34 @@ class Post(models.Model):
     text = models.TextField()
     created_by = models.ForeignKey("auth.User", on_delete=models.CASCADE)
     creation_datetime = models.DateTimeField(auto_now_add=True)
-    # like_count = models.IntegerField(default=0)
     tags = models.ManyToManyField(Tag, help_text="Ctrl + click to select multiple", blank=True)
 
     def __str__(self):
         return self.created_by.username
+
+    @property
+    def like_count(self):
+        return self.likes.count()
+
+    @property
+    def average_rate(self):
+        return self.rates.aggregate(Avg('value'))['value__avg']
+
+
+class Like(models.Model):
+    post = models.ForeignKey("Post", on_delete=models.CASCADE, related_name="likes")
+    user = models.ForeignKey("auth.User", on_delete=models.CASCADE)
+
+
+class Rate(models.Model):
+    value = models.PositiveSmallIntegerField(choices=[(i, i) for i in [0, 1, 2, 3, 4, 5]], default=0)
+    post = models.ForeignKey("Post", on_delete=models.CASCADE, related_name="rates")
+    user = models.ForeignKey("auth.User", on_delete=models.CASCADE)
+
+    def save(self, *args, **kwargs):
+        self.full_clean()
+        super().save(*args, **kwargs)
+
+    def clean(self):
+        if not Rate.objects.filter(post=self.post).exists():
```

Comment:
This doesn't check user. It should be `Rate.objects.filter(post=self.post, user=self.user).exists()`. Also it should be in `if self._state_adding:` block, otherwise, it will raise error on updates too.

However, there is a better solution for this case, which is using `Unique Constraint` to prevent creating multiple rates with the same post&user couple. Sorry for misleading to `clean` approach at first, but that's an experience too.

Reply:
Summary, delete save and clean overrides and use `Unique Constraint`.